src/S3Persistor.js

```diff
@@ -0,0 +1,366 @@
+const http = require('http')
+const https = require('https')
+if (http.globalAgent.maxSockets < 300) {
+  http.globalAgent.maxSockets = 300
+}
+if (https.globalAgent.maxSockets < 300) {
+  https.globalAgent.maxSockets = 300
+}
+
+const Settings = require('./Settings')
+
+const PersistorHelper = require('./PersistorHelper')
+
+const fs = require('fs')
+const S3 = require('aws-sdk/clients/s3')
+const { URL } = require('url')
+const Stream = require('stream')
+const { promisify, callbackify } = require('util')
+const {
+  WriteError,
+  ReadError,
+  NotFoundError,
+  SettingsError
+} = require('./Errors')
+const pipeline = promisify(Stream.pipeline)
+
+const S3Persistor = {
+  sendFile: callbackify(sendFile),
+  sendStream: callbackify(sendStream),
+  getObjectStream: callbackify(getObjectStream),
+  getRedirectUrl: callbackify(getRedirectUrl),
+  getObjectMd5Hash: callbackify(getObjectMd5Hash),
+  deleteDirectory: callbackify(deleteDirectory),
+  getObjectSize: callbackify(getObjectSize),
+  deleteObject: callbackify(deleteObject),
+  copyObject: callbackify(copyObject),
+  checkIfObjectExists: callbackify(checkIfObjectExists),
+  directorySize: callbackify(directorySize),
+  promises: {
+    sendFile,
+    sendStream,
+    getObjectStream,
+    getRedirectUrl,
+    getObjectMd5Hash,
+    deleteDirectory,
+    getObjectSize,
+    deleteObject,
+    copyObject,
+    checkIfObjectExists,
+    directorySize
+  }
+}
+
+module.exports = S3Persistor
+
+async function sendFile(bucketName, key, fsPath) {
+  return sendStream(bucketName, key, fs.createReadStream(fsPath))
+}
+
+async function sendStream(bucketName, key, readStream, sourceMd5) {
+  try {
+    // egress from us to S3
+    const observeOptions = { metric: 's3.egress' }
+    let b64Hash
+
+    if (sourceMd5) {
+      b64Hash = PersistorHelper.hexToBase64(sourceMd5)
+    } else {
+      // if there is no supplied md5 hash, we calculate the hash as the data passes through
+      observeOptions.hash = 'md5'
+    }
+
+    const observer = new PersistorHelper.ObserverStream(observeOptions)
+    pipeline(readStream, observer)
+
+    // if we have an md5 hash, pass this to S3 to verify the upload
+    const uploadOptions = {
+      Bucket: bucketName,
+      Key: key,
+      Body: observer
+    }
+    if (b64Hash) {
+      uploadOptions.ContentMD5 = b64Hash
+    }
+
+    const response = await _getClientForBucket(bucketName)
+      .upload(uploadOptions, { partSize: Settings.s3.partSize })
+      .promise()
+    let destMd5 = _md5FromResponse(response)
+    if (!destMd5) {
+      // the eTag isn't in md5 format so we need to calculate it ourselves
+      const verifyStream = await getObjectStream(
+        response.Bucket,
+        response.Key,
+        {}
+      )
```

Comment:
Oh wow! ok. So that's what you warned about in the README. We need to download the file that we just uploaded to S3. Does that even work all the time? According to [the S3 docs](https://docs.aws.amazon.com/AmazonS3/latest/dev/Introduction.html#CoreConcepts), "Amazon S3 provides read-after-write consistency for PUTs of new objects in your S3 bucket", but not for ovewrite PUTs. So, is it possible that we might incorrectly reject a file update?

Reply:
> read-after-write consistency

Ah, right yes. Filestore deletes what's there before overwriting but we're not doing that here.... 

It looks like we can supply a `computeChecksums` option to the S3 upload. This means that we only have to validate the data locally if the client has provided an md5 hash as it passes through, as the S3 client will compute the checksums to validate the "local" data (relative to this module) matches the data that S3 receives.

Does that sound right to you?

Reply:
https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#constructor-property

Reply:
That looks good. If I understand correctly, the `ManagedUpload` object buffers the streams in chunks and calculates the MD5 on each chunk before sending it.